Add suggestions to the interchain exchange tutorial (#575)
cosmos/sdk-tutorials: interchain-exchange/tutorial/05-order-initialization.md

---
order: 6
---

# Initialize the Orders

> In this chapter you will initialize the bank keeper and the modules. 

Reviewer: > you will initialize the bank keeper

We rather import bank keeper method into `ibcdex`

> You will add the bank module initialization to the blockchain app that you have created so you can use it in the exchange module. 

Reviewer: Bank is already initialized by default

> In this chapter you will make sure that there can be no new native tokens created on a blockchain, which will secure the exchange module from any blockchain misbehaviour.

Reviewer: ```suggestion
In this chapter, you will make sure that there can be no new native tokens minted on a blockchain, which will secure the exchange module from any counterparty misbehavior.
```

## Add the bank keeper to the module

> Add the keeper to the `ibcdex` module. First, in the `types` directory.

Reviewer: ```suggestion
Add the bank keeper reference to the `ibcdex` module. First, in the `types` directory.
```

```go
// app/app.go
app.ibcdexKeeper = *ibcdexkeeper.NewKeeper(
  appCodec,
	keys[ibcdextypes.StoreKey],
	keys[ibcdextypes.MemStoreKey],
	app.IBCKeeper.ChannelKeeper,
	&app.IBCKeeper.PortKeeper,
	scopedIbcdexKeeper,
	app.BankKeeper,
)
```

The app will mint and burn token using the bank account. The use this feature, the module must have a _module account_. To enable the _module account_ you declare this permission in the _module account permissions_ structure of the auth module. This is done in `app.go` file.

```go
// app/app.go
maccPerms = map[string][]string{
    authtypes.FeeCollectorName:     nil,
    distrtypes.ModuleName:          nil,
    minttypes.ModuleName:           {authtypes.Minter},
    stakingtypes.BondedPoolName:    {authtypes.Burner, authtypes.Staking},
    stakingtypes.NotBondedPoolName: {authtypes.Burner, authtypes.Staking},
    govtypes.ModuleName:            {authtypes.Burner},
    ibctransfertypes.ModuleName:    {authtypes.Minter, authtypes.Burner},
    ibcdextypes.ModuleName: 		{authtypes.Minter, authtypes.Burner}, // <--
}
```

> ## Ensure No Native Token are Minted

Reviewer: ```suggestion
## Ensure no native token is minted
```
Mmh... why uppercase everywhere?

> Ensure that no new native tokens are minted in case the module is connected to a faulty blockchain. Rename this functionality into lock and unlock and reuse functions from `ibc-transfer` module with an escrow address to ensure no new token creation.

Reviewer: ```suggestion
Ensure that no new native tokens are minted in case the module is connected to a malicious or faulty counterparty. Rename this functionality into lock and unlock and reuse functions from `ibc-transfer` module with an escrow address to ensure no new token creation.
```

For non-native tokens, use the _module accounts_ functionality.

> To abstract the process of deciding if a token must be locked or burned, introduce `SafeBurn` and `SafeMint`. Those functions will check if the denom is native to the chain or a voucher.

Reviewer: ```suggestion
To abstract the process of deciding if a token must be locked or burned, we introduce `SafeBurn` and `SafeMint`. Those functions will check if the denom is native to the chain or a voucher.
```

> Consider that any denom staring with `ibc/` is a voucher.

Reviewer: ```suggestion
Consider that any denom starting with `ibc/` is a voucher.
```

		ctx, types.ModuleName, receiver, sdk.NewCoins(tokens),
	); err != nil {
		panic(fmt.Sprintf("unable to send coins from module to account despite previously minting coins to module account: %v", err))
	}

	return nil
}

func (k Keeper) LockTokens(
	ctx sdk.Context,
	sourcePort string,
	sourceChannel string,
	sender sdk.AccAddress,
	tokens sdk.Coin,
) error {
	// create the escrow address for the tokens
	escrowAddress := ibctransfertypes.GetEscrowAddress(sourcePort, sourceChannel)

	// escrow source tokens. It fails if balance insufficient
	if err := k.bankKeeper.SendCoins(
		ctx, sender, escrowAddress, sdk.NewCoins(tokens),
	); err != nil {
		return err
	}

	return nil
}

func (k Keeper) UnlockTokens(
	ctx sdk.Context,
	sourcePort string,
	sourceChannel string,
	receiver sdk.AccAddress,
	tokens sdk.Coin,
) error {
	// create the escrow address for the tokens
	escrowAddress := ibctransfertypes.GetEscrowAddress(sourcePort, sourceChannel)

	// escrow source tokens. It fails if balance insufficient
	if err := k.bankKeeper.SendCoins(
		ctx, escrowAddress, receiver, sdk.NewCoins(tokens),
	); err != nil {
		return err
	}

	return nil
}
```

> ## Declare the Denom

Reviewer: ```suggestion
## Declare the denom
```

> The token denoms should have the same behavior as the `ibc-transfer` module.

Reviewer: I would say we should remove this statement. People would think the denom would be compatible with `ibc-transfer` while this is not the case. We have here a simplified system to manage voucher denom different from `ibc-transfer`

> - An external token received from a chain has a unique denom, reffered to as `voucher`

Reviewer: ```suggestion
- An external token received from a chain has a unique denom, referred to as `voucher`
```

- When a token sent to a chain is received back, the chain can resolve the voucher and convert it back to the original token denomination

Vouchers are hashes, therefore you must store which original denomination is related to a voucher, do this with an `indexed` type transaction.

> For a voucher store: 

Reviewer: ```suggestion
For a voucher, store: 
```
